ComponentKit/Utilities/CKArrayControllerChangesetVerification.mm

```diff
@@ -0,0 +1,158 @@
+    }];
+  }
+
+  /** 
+   Removed sections
+   
+   We need to keep an index set to track what sections to remove instead of removing each inline because we cannot
+   guarantee the order of the removals, so for example:
+   
+   Let's say we have 3 sections, and we want to remove the second and third section. If we remove the second section first,
+   we'll end up trying to remove the third section when there's only 2 sections left, and we'll get a false negative.
+   */
+  NSMutableIndexSet *sectionsToRemove = [NSMutableIndexSet indexSet];
+  for (NSInteger removal : changeset.sections.removals()) {
+    if (removal >= [currentSectionCounts count]) {
+      return NO;
+    } else {
+      [sectionsToRemove addIndex:removal];
+    }
+  }
+  [currentSectionCounts removeObjectsAtIndexes:sectionsToRemove];
+
+  // Inserted sections
+  for (NSInteger insertion : changeset.sections.insertions()) {
+    // Strictly greater than, since you can always insert at the end
+    if (insertion > [currentSectionCounts count]) {
+      return NO;
+    } else {
+      [currentSectionCounts insertObject:@0 atIndex:insertion];
+    }
+  }
+
+  // Moved sections
+  for (std::pair<NSInteger, NSInteger> move : changeset.sections.moves()) {
+    NSInteger origin = move.first;
+    NSInteger destination = move.second;
+
+    if (origin >= [currentSectionCounts count] || destination >= [currentSectionCounts count]) {
+      return NO;
+    } else {
+      NSNumber *originValue = currentSectionCounts[origin];
+      NSNumber *destinationValue = currentSectionCounts[destination];
+
+      [currentSectionCounts replaceObjectAtIndex:origin withObject:destinationValue];
+      [currentSectionCounts replaceObjectAtIndex:destination withObject:originValue];
+    }
+  }
+
+  // Inserted rows
+  changeset.items.enumerateItems(nil, nil, ^(NSInteger section, NSInteger index, id<NSObject> object, BOOL *stop) {
+    if (section >= [currentSectionCounts count] || index > [currentSectionCounts[section] integerValue]) {
+      invalidChangeFound = YES;
+      *stop = YES;
+    } else {
+      currentSectionCounts[section] = @([currentSectionCounts[section] integerValue] + 1);
+    }
+  }, nil);
+
+  if (invalidChangeFound) {
+    return NO;
+  }
+
+  // Moved rows
+  changeset.items.enumerateItems(nil, nil, nil, ^(const CKArrayControllerIndexPath &fromIndexPath, const CKArrayControllerIndexPath &toIndexPath, BOOL *stop) {
+    BOOL fromIndexPathSectionInvalid = fromIndexPath.section >= [currentSectionCounts count];
+    BOOL toIndexPathSectionInvalid = toIndexPath.section >= [currentSectionCounts count];
+
+    /**
+     First only check section validity.
+     If we don't do this first, we risk an index out of bounds crash in the item validity check
+     */
+    if (fromIndexPathSectionInvalid ||
+        toIndexPathSectionInvalid) {
+      invalidChangeFound = YES;
+      *stop = YES;
+      // Need this return so that it doesn't continue to evaluate
+      return;
```

Comment:
This is a bit fragile. Maybe break this block up into a couple helper functions and add an else condition below (line 140)? That way we don't need to explicitly return here.